presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -352,4 +472,316 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
+                columnIndex.getNullCounts() == null || columnIndex.getNullPages() == null) {
+            return true;
+        }
+
+        if (columnIndex.getMaxValues().size() != columnIndex.getMinValues().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullPages().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullCounts().size()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    // Caller should verify isCorruptedColumnIndex is false first
+    private boolean isEmptyColumnIndex(ColumnIndex columnIndex)
+    {
+        return columnIndex.getMaxValues().size() == 0;
+    }
+
+    public FilterPredicate convertToParquetUdp()
+    {
+        FilterPredicate filter = null;
+
+        //  It could be a Presto bug that we don't see effectivePredicate.getDomains().get() has more than 1 domain.
+        //  For example, the 'where c1=3 or c1=10002' clause should have two domains but it has none
+        //  we assume the relation cross domains are 'or'
+        for (RichColumnDescriptor column : columns) {
+            Domain domain = effectivePredicate.getDomains().get().get(column);
+            if (domain == null || domain.isNone()) {
+                continue;
+            }
+
+            if (domain.isAll()) {
+                continue;
+            }
+
+            FilterPredicate columnFilter = FilterApi.userDefined(FilterApi.intColumn(ColumnPath.get(column.getPath()).toDotString()), new DomainUserDefinedPredicate(domain));
+            if (filter == null) {
+                filter = columnFilter;
+            }
+            else {
+                filter = FilterApi.or(filter, columnFilter);
+            }
+        }
+
+        return filter;
+    }
+
+    /**
+     * This class implements methods defined in UserDefinedPredicate based on the page statistic and tuple domain(for a column).
+     */
+    static class DomainUserDefinedPredicate<T extends Comparable<T>>
+            extends UserDefinedPredicate<T>
+            implements Serializable
+    {
+        private Domain columnDomain;
+
+        DomainUserDefinedPredicate(Domain domain)
+        {
+            this.columnDomain = domain;
+        }
+
+        @Override
+        public boolean keep(T value)
+        {
+            if (value == null && !columnDomain.isNullAllowed()) {
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public boolean canDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistic)
+        {
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
+                    Integer min = (Integer) statistic.getMin();
+                    Integer max = (Integer) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics((long) min, (long) max));
```

Comment:
is canDrop defined in parquet apis?
does it mean we could safely skip the pages based on statistics?

Reply:
Yea, it is an interface method and we use it to drop when true.